[NTOS:PS] Basic support for jobs in the kernel: Create/Open/Assign/Remove/Terminate/Delete

wnsntgs opened the pull request:
## Abstract

Jobs are kernel objects that allow processes to be grouped and manipulated as a single entity (for example, in order to apply constraints to them and the threads they contain). If a process is in a job, then all processes’ threads that those processes create will also be in the job, with no escape.

## Background

This PR supersedes the old WIP PR #40 by  Mark Jansen and is partially based on it (which in turn was based on the patch by Samuel Serapión Vega). Moreover, the implementation of `JobObjectBasicLimitInformation` and `JobObjectExtendedLimitInformation` limits in `NtSetInformationJobObject()` is based on  Timo Kreuzer's patch.

This PR adds basic support for pre-Windows 8 job objects to the kernel. It adapts the previous patch part-by-part, gradually adding to its unfinished parts as well fixing some things (like handling job status flags or fixing input validation) because the patch did not produce satisfactory results when running WINE tests, and adds more functionality on top of it. 

This is not the final version of what functional job support should look like; some features are intentionally left untouched in this PR.

## Current status

- Creating a job object (`NtCreateJobObject`)
- Opening a job object (`NtOpenJobObject`)
- Adding a process to a job (`NtAssignProcessToJobObject`, `PspAssignProcessToJob`)
- Removing a process from a job (`PspRemoveProcessFromJob`)
- Exiting a process from a job (`PspExitProcessFromJob`)
- Checking if a process is in the job (`NtIsProcessInJob`)
- Terminating processes in a job (`NtTerminateJobObject`, `PspTerminateJobObject`)
- Enumeration of processes in a job (`PspEnumerateProcessesInJob`)
- Closing a job object (`PspCloseJob`)
- Deleting a job object (`PspDeleteJob`)
- Querying job object information (`NtQueryInformationJobObject`)
  - `JobObjectBasicAccountingInformation`, `JobObjectBasicAndIoAccountingInformation`, `JobObjectBasicLimitInformation`, `JobObjectExtendedLimitInformation`, `JobObjectBasicProcessIdList`
- Setting job object information (`NtSetInformationJobObject`)
  - `JobObjectBasicLimitInformation`, `JobObjectExtendedLimitInformation`, `JobObjectAssociateCompletionPortInformation`

Job-related tests from the WINE kernel32 test suite:

<details>
  <summary>Before (60 failures):</summary>

  ![kernel32_winetest process - before](https://imgur.com/A6ulmtm.jpg)
</details>

<details>
  <summary>After (0 failures):</summary>

  ![kernel32_winetest process - after](https://i.imgur.com/yQkbOB2.png)
</details>

## TODO

Plans for future commits/PRs:
- Add support for job limits (support more job information classes in `NtQueryInformationJobObject`/`NtSetInformationJobObject`).
- Add support for job sets.
- Figure out and support more job/job status flags.
- 100% pass rate of job related tests in the WINE kernel32 test suite. This implies:
  - [x] Support for associating a completion port with a job.
  - [x] Fixing breakaway scenarios.

wnsntgs committed:
[NTOS:PS] As part of improving job object kernel functionality:
- Clean up NtCreateJobObject
- Modernize its SAL annotations
- Add a documenting comment for NtCreateJobObject

wnsntgs committed:
[NTOS:PS] NtOpenJobObject:
- SAL2 annotate NtOpenJobObject and related user-mode functions
- Re-format NtOpenJobObject
- Don't use hungarian notation for handles in NtCreateJobObject and NtOpenJobObject

wnsntgs committed:
[NTOS:PS] Extend pstypes with jobs related types
>
>
Co-authored-by: Mark Jansen <mark.jansen@reactos.org>
Co-authored-by: Samuel Serapión Vega <encoded@reactos.org>

wnsntgs committed:
[NTOS:PS] Basic support for assigning a process to a job, WIP
>
>
Co-authored-by: Mark Jansen <mark.jansen@reactos.org>
Co-authored-by: Samuel Serapión Vega <encoded@reactos.org>

wnsntgs committed:
[NTOS:PS] Handle JOB_OBJECT_MSG_ACTIVE_PROCESS_LIMIT in PspAssignProcessToJob

wnsntgs committed:
[NTOS:PS] NtIsProcessInJob/IsProcessInJob:
- Modernize SAL annotations
- Handle the case when the passed ProcessHandle is the current process
- Add comments

wnsntgs committed:
[NTOS:PS] Change the order of functions in the file to follow the logical principle:
(1) Nt functions come after Ps.
(2) Initialization and creation functions first, then open, assign, terminate, delete, and so on down to getters and setters.

wnsntgs committed:
[NTOS:PS] Implement basic PspTerminateJobObject

Co-authored-by: Mark Jansen <mark.jansen@reactos.org>
Co-authored-by: Samuel Serapión Vega <encoded@reactos.org>

wnsntgs committed:
[NTOS:PS] Get rid of ambiguous characters in pstypes

wnsntgs committed:
[NTOS:PS] Implement PspRemoveProcessFromJob

Co-authored-by: Mark Jansen <mark.jansen@reactos.org>
Co-authored-by: Samuel Serapión Vega <encoded@reactos.org>

wnsntgs committed:
[NTOS:PS] Implement PspExitProcessFromJob

Co-authored-by: Mark Jansen <mark.jansen@reactos.org>
Co-authored-by: Samuel Serapión Vega <encoded@reactos.org>

wnsntgs committed:
[NTOS:PS] Implement PspCloseJob and extend the job type object initializer

Co-authored-by: Mark Jansen <mark.jansen@reactos.org>
Co-authored-by: Samuel Serapión Vega <encoded@reactos.org>

wnsntgs committed:
[NTOS:PS] Fix handling of the CREATE_BREAKAWAY_FROM_JOB flag in CreateProcessInternalW

Co-authored-by: Mark Jansen <mark.jansen@reactos.org>

wnsntgs committed:
[NTOS:PS] NtCreateJobObject: Initialize SchedulingClass during job object creation and revise comments

wnsntgs committed:
[NTOS:PS] Add a general-purpose process enumeration function that can be used in other places where processes need to be enumerated within a job object
And:
- Refactor PspTerminateJobObject to use a process enumeration callback

wnsntgs committed:
[NTOS:PS] Add PspEnumerateProcessesInJob to ps.h

wnsntgs committed:
[NTOS:PS] Implement job assignment when creating a process
And:
- Add PspAssignProcessToJob to ps.h

Co-authored-by: Samuel Serapión Vega <encoded@reactos.org>

wnsntgs committed:
[NTOS:PS] Fix a false positive assertion failure
And:
- Minor style fixes (clean up excessive comments, respect 80 cols)

wnsntgs committed:
[NTOS:PS] Dereference job object in NtAssignProcessToJobObject twice when it has been referenced twice already

wnsntgs committed:
[NTOS:PS] Bug fixes:
- Use JobStatus flags (which starting with Windows Vista are in EPROCESS::Flags2 and have names). This allows us to replace the magic number in the old job code and fix the problem in the new code detected by the wine kernel32 tests. In the case of job termination, a situation may arise where the number of active processes may be decremented twice. The problem is solved by using the job inactivity flag, which is set atomically in the decrementing code. Think: terminate_job -> decrement -> terminate_process -> exit_last_thread -> remove_from_job -> decrement

wnsntgs committed:
[NTOS:PS] Fix job enumeration
And:
- Don't try to set the job object to NULL too early in PspRemoveProcessFromJob as the process destructor may get NULL

wnsntgs committed:
[NTOS:PS] Prevent processes from being added to a job if it is flagged for closing and has a limit on process termination on closing
And:
- Add JOB_OBJECT_CLOSE_DONE to pstypes.h, name comes from the JobFlags union with bit fields which was introduced in Windows 8

wnsntgs committed:
[NTOS:PS] Move internal job flags to ps.h

wnsntgs committed:
[NTOS:PS] If the job is terminating when AssignProcessToJob is called, the function should fail

wnsntgs committed:
[NTOS:PS] Implement JobObjectBasicLimitInformation and JobObjectExtendedLimitInformation limits in NtSetInformationJobObject based on the patch by Timo Kreuzer
And:
- Separate this functionality into a helper function
- Modernize SAL annotations

Co-authored-by: Timo Kreuzer <timo.kreuzer@reactos.org>

wnsntgs committed:
[NTOS:PS] Add privilege checking when setting JOB_OBJECT_LIMIT_PRIORITY_CLASS or JOB_OBJECT_LIMIT_SCHEDULING_CLASS job limits as per MicrosoftDocs
And:
- Clarify comments in NtSetInformationJobObject

wnsntgs committed:
[NTOS:PS] Add parameter/privilege validation for JOB_OBJECT_LIMIT_WORKINGSET and JOB_OBJECT_LIMIT_AFFINITY limit flags

wnsntgs committed:
[NTOS:PS] Implement JobObjectBasicProcessIdList
And:
- Document and refactor existing NtQueryInformationJobObject code using helper functions

wnsntgs committed:
[NTOS:PS] Improve job enumeration and simplify some code
And:
- Add the ability to exit job enumeration early on first failure
- Add more detailed comment describing the purpose of the jobs file

wnsntgs committed:
[NTOS:PS] Fix race conditions and improve synchronization in job object termination
- Correct improper use of InterlockedOr when clearing JOB_OBJECT_TERMINATING flag, preventing unintended flag modifications
- Move job flag and process state modifications inside critical sections
- Protect existing job completion signaling to ensure consistency

wnsntgs committed:
[NTOS:PS] Refactor PspQueryJobProcessIdList to use callback-based process enumeration

wnsntgs committed:
[NTOS:PS] Fix for Status might not be initialized

wnsntgs committed:
[NTOS:PS] Fix returns without releasing job lock

wnsntgs committed:
[NTOS:PS] Fix some comments

UIRestrictionsClass should (and can) only be updated for a job

wnsntgs committed:
[NTOS:PS] Implement JobObjectAssociateCompletionPortInformation

wnsntgs committed:
[NTOS:PS] Address FIXME in ps\kill.c, move JOB_OBJECT_MSG_EXIT_PROCESS to the right place

This fixes an unexpected overlapped in the WINE test.

wnsntgs committed:
[NTOS:PS] Fix compilation errors

DarkFire01 commented:
PR part of the bundle that allowed firefox 75 to work!
![image](https://github.com/user-attachments/assets/71989e07-634f-4dd8-aa66-2316ffe4533d)

learn-more reviewed (changes requested):
I think the reference counting could do with another pass over it,
other than that a lot of the stuff looks good.

wnsntgs committed:
[NTOS:PS] Remove logically impossible condition (JOB_OBJECT_LIMIT_WORKINGSET)

wnsntgs committed:
[NTOS:PS] Fix extra dereference of the job object in NtAssignProcessToJobObject

wnsntgs committed:
[NTOS:PS] Ensure that if the caller passed a handle that is not the current process, proper dereferencing is performed

wnsntgs committed:
[NTOS:PS] Acquire job lock exclusively during job limit modifications

wnsntgs committed:
[NTOS:PS] Print out the invalid LimitFlags when an unsupported flag is detected

wnsntgs committed:
[NTOS:PS] Dereference the completion port before setting the pointer to NULL

wnsntgs committed:
[NTOS:PS] Add debug logging for unexpected SystemHandleCount in PspJobClose

wnsntgs committed:
[NTOS:PS] Notify the completion port in PspRemoveProcessFromJob

wnsntgs committed:
[NTOS:PS] Fix race condition in termination flag

Set the JOB_OBJECT_TERMINATING flag once before enumeration and clear after all processes are terminated instead of setting/clearing it per-process.

wnsntgs committed:
[NTOS:PS] Release the proces reference before breaking in PspEnumerateProcessesInJob

wnstngs commented on ntoskrnl/ps/job.c:
```diff
@@ -80,159 +84,1413 @@ ULONG PspJobInfoAlign[] =
+    /* Proceed only when the last handle is left */
+    if (SystemHandleCount != 1)
+    {
+        return;
```
Done.

wnstngs commented on ntoskrnl/ps/job.c:
```diff
@@ -80,159 +84,1413 @@ ULONG PspJobInfoAlign[] =
+    }
+
+    /* Set the completion port to NULL to clean up */
+    Job->CompletionPort = NULL;
```
Thanks, fixed.

wnstngs commented on ntoskrnl/ps/job.c:
```diff
@@ -80,159 +84,1413 @@ ULONG PspJobInfoAlign[] =
+    /* Validate flags */
+    if (ExtendedLimit->BasicLimitInformation.LimitFlags & ~AllowedFlags)
+    {
+        DPRINT1("Invalid LimitFlags specified\n");
```
Done.

wnstngs commented on ntoskrnl/ps/job.c:
```diff
@@ -80,159 +84,1413 @@ ULONG PspJobInfoAlign[] =
+    }
+
+    /* Acquire the job lock */
+    ExAcquireResourceSharedLite(&Job->JobLock, TRUE);
```
You're right, switched to exclusive and APCs are already disabled before this point.

wnstngs commented on ntoskrnl/ps/job.c:
```diff
@@ -80,159 +84,1413 @@ ULONG PspJobInfoAlign[] =
+                ExtendedLimit->BasicLimitInformation.MinimumWorkingSetSize)
+            ||
+            (!ExtendedLimit->BasicLimitInformation.MaximumWorkingSetSize &&
+                ExtendedLimit->BasicLimitInformation.MaximumWorkingSetSize))
```
Thanks, fixed.

wnstngs commented on ntoskrnl/ps/job.c:
```diff
@@ -80,159 +84,1413 @@ ULONG PspJobInfoAlign[] =
+    Context.CompletionPort = Job->CompletionPort;
+    Context.CompletionKey = Job->CompletionKey;
+
+    Status = PspEnumerateProcessesInJob(Job,
```
Done.

wnstngs commented on ntoskrnl/ps/job.c:
```diff
@@ -283,270 +1554,395 @@ NtCreateJobObject (
+    }
+
+    /* Dereference the process object if it was referenced */
+    if (ProcessHandle == PsGetCurrentProcess())
```
Thanks, fixed.

wnstngs commented on ntoskrnl/ps/job.c:
```diff
@@ -80,159 +84,1413 @@ ULONG PspJobInfoAlign[] =
+            AnyCallbackFailed = TRUE;
+            if (BreakOnCallbackFailure)
+            {
+                break;
```
Thanks, added dereference.

wnstngs commented on ntoskrnl/ps/job.c:
```diff
@@ -80,159 +84,1413 @@ ULONG PspJobInfoAlign[] =
+           processes from repeatedly decrementing */
+        InterlockedOr((PLONG)&Process->JobStatus, JOB_NOT_REALLY_ACTIVE);
+    }
+
```
That's right, thanks - added it.

wnsntgs committed:
[NTOS:PS] Fix C4101

wnsntgs committed:
[NTOS:PS] Always release the job lock before PspAssignProcessToJob returns

wnsntgs committed:
[NTOS:PS] Return the correct status in PspAssignProcessToJob if the job has JOB_OBJECT_LIMIT_ACTIVE_PROCESS limit

wnsntgs committed:
[NTOS:PS] Balance reference management in process assignment so that the job object remains alive for all subsequent operations

wnsntgs committed:
[NTOS:PS] Fix crashes

wnsntgs committed:
[NTOS:PS] Use the job object itself instead of additional context structure when associating an IOCP via a callback

wnsntgs committed:
[NTOS:PS] Call IoSetIoCompletion in PspTerminateProcessCallback if there are no active processes left in the job

wnstngs commented on ntoskrnl/ps/job.c:
```diff
@@ -283,270 +1554,395 @@ NtCreateJobObject (
-            ObDereferenceObject(Job);
-            return Status;
+            /* At this point, the job was referenced twice */
+            ObDereferenceObjectEx(Job, 2);
```
It looks correct now.

wnsntgs committed:
[NTOS:PS] Terminate the process if the assignment causes the active process count to exceed ActiveProcessLimit as per the JOBOBJECT_BASIC_LIMIT_INFORMATION documentation(*)

(*) - https://learn.microsoft.com/en-us/windows/win32/api/winnt/ns-winnt-jobobject_basic_limit_information

tkreuzer commented on ntoskrnl/ps/job.c:
```diff
@@ -80,159 +84,1413 @@ ULONG PspJobInfoAlign[] =
+    if (Job->LimitFlags & JOB_OBJECT_LIMIT_ACTIVE_PROCESS)
+    {
+        /* Check if job limit on active processes has been reached */
+        if (Job->ActiveProcesses > Job->ActiveProcessLimit)
```
```suggestion
        if (Job->ActiveProcesses => Job->ActiveProcessLimit)
```

tkreuzer commented on ntoskrnl/ps/job.c:
```diff
@@ -80,159 +84,1413 @@ ULONG PspJobInfoAlign[] =
+       when AssignProcessToJob is called, the function fails" */
+    if (Job->JobFlags & JOB_OBJECT_TERMINATING)
+    {
+        return STATUS_INVALID_PARAMETER;
```
Leaking the lock here and below.

```suggestion
        Status = STATUS_INVALID_PARAMETER;
        goto Exit;
```

tkreuzer reviewed (commented):
Can we get this completed?

wnsntgs committed:
[NTOS:PS] Improve job management

Introduce a common lock held iterator for direct job membership

Make process assignment a single transaction protected by process rundown
and per-job lock

Atomically publish Process->Job, maintain the process owned job
ref, and update membership and counters without exposing partially
committed assignment state

Improve job termination bookkeeping by making active-to-inactive transitions
idempotent and preventing duplicate notifications

Serialize completion port lifetime against assignment, exit and close

Also improve job close cleanup, synchronization assertions and documentation

wnsntgs commented on ntoskrnl/ps/job.c:
```diff
@@ -80,159 +84,1413 @@ ULONG PspJobInfoAlign[] =
+    ExEnterCriticalRegionAndAcquireResourceExclusive(&Job->JobLock);
+
+    /* Check if the process is part of the specified job */
+    if (Process->Job == Job)
```
Agreed, now PspRemoveProcessFromJob and PspExitProcessFromJob only accept a process as a parameter.

learn-more commented on ntoskrnl/ps/job.c:
```diff
@@ -74,479 +78,2154 @@ ULONG PspJobInfoAlign[] =
+
+    /* Zero the basic accounting information */
+    RtlZeroMemory(&BasicAndIo->BasicInfo,
+                  sizeof(JOBOBJECT_BASIC_ACCOUNTING_INFORMATION));
```
nit: could do `sizeof(BasicAndIo->BasicInfo)`

wnsntgs replied:
Done here and in a few other places

learn-more commented on ntoskrnl/ps/job.c:
```diff
@@ -74,479 +78,2154 @@ ULONG PspJobInfoAlign[] =
+#endif
+
+    /* Ensure the process is active and has a valid unique process ID */
+    if (!(Process->JobStatus & JOB_NOT_REALLY_ACTIVE) &&
```
Might as well use the `BooleanFlagOn` family of functions?

wnsntgs replied:
I didn't know that, switched to using BooleanFlagOn where it made sense

tkreuzer replied:
Just FYI: `FlagOn` would be enough here. `BooleanFlagOn` is only to guarantee that the result is either `TRUE` or `FALSE`, e.g. when you want to assign it to a `BOOLEAN`.

learn-more commented on ntoskrnl/ps/job.c:
else, should it be cleaned up?

wnsntgs replied:
ObInsertObject consumes the ref returned by ObCreateObject and dereferences the object on failure, so PspDeleteJob performs the cleanup

tkreuzer commented on ntoskrnl/ps/job.c:
```diff
@@ -74,479 +78,2154 @@ ULONG PspJobInfoAlign[] =
+{
+    ASSERT(Iterator != NULL);
+
+#if DBG
```
The `#if DBG` is not needed here and in other cases below.

wnsntgs replied:
Removed

tkreuzer commented on ntoskrnl/ps/job.c:
```diff
@@ -74,479 +78,2154 @@ ULONG PspJobInfoAlign[] =
+typedef struct PSP_JOB_ITERATOR
+{
+    PEJOB Job;
+    PLIST_ENTRY NextEntry;
```
I would recommend to use a `PEPROCESS` here, which is more convenient for debugging.

wnsntgs replied:
Removed PSP_JOB_ITERATOR

tkreuzer commented on ntoskrnl/ps/job.c:
```diff
@@ -74,479 +78,2154 @@ ULONG PspJobInfoAlign[] =
+ */
+static
+PEPROCESS
+PspAdvanceJobEnumerator(
```
Is the discrepency in the naming (`JobEnumerator` vs `JobIterator`) intended?

wnsntgs replied:
This is no longer relevant, simplified the implementation. The point of making it overengineered was to prepare for Windows 8+ jobs, where they can be nested, making it more difficult to enumerate them. I was trying to separate: a locked iterator returning borrowed processes and an unlocked enumerator returning referenced processes, with the idea that the iterator state could later grow into something suitable for nesting.

tkreuzer commented on ntoskrnl/ps/job.c:
```diff
@@ -74,479 +78,2154 @@ ULONG PspJobInfoAlign[] =
+
+    PspInitializeJobIteratorLocked(Job, Process, &Iterator);
+
+    while ((Candidate = PspAdvanceJobIteratorLocked(&Iterator)) != NULL)
```
All of the code above this function could be replaced with something like
```
    ListLink = Process ? Process ->JobLinks.Flink : Job->ProcessListHead.Flink;
    for ( ; ListLink != &Job->ProcessListHead; ListLink = ListLink->Flink)
    {
        Candidate = CONTAINING_RECORD(ListLink, EPROCESS, JobLinks);
        ...
```
In other words: it seems a bit overengineered to abstract a simple linked list iteration with an iterator structure and 2 functions.
Alternatively something as simple as
```
static PEPROCESS PspGetNextProcessInJob(PEJOB Job, PEPROCESS Process)
{
    PLIST_ENTRY ListLink = Process ? Process ->JobLinks.Flink : Job->ProcessListHead.Flink;
    if (ListLink == &Job->ProcessListHead)
        return NULL;
    return CONTAINING_RECORD(ListLink, EPROCESS, JobLinks);
}
```
would also do it, make the loop a bit more elegant and avoid all the clutter.

wnsntgs replied:
Simplified the enumeration implementation

tkreuzer commented on ntoskrnl/ps/job.c:
```diff
@@ -74,479 +78,2154 @@ ULONG PspJobInfoAlign[] =
+    Process = PspAdvanceJobEnumerator(Job, NULL);
+
+    /* Iterate through all processes in the job */
+    while (Process)
```
Alternatively
```
    Process = NULL;
    while ((Process = PspAdvanceJobEnumerator(Job, Process)) != NULL)
```
Like it's done above and below.

tkreuzer commented on ntoskrnl/ps/job.c:
```diff
@@ -74,479 +78,2154 @@ ULONG PspJobInfoAlign[] =
+
+    /* Acquire the reference owned by the process for as long as
+       Process->Job points to Job */
+    ObReferenceObject(Job);
```
You could do that below after successfully assigning it.

wnsntgs replied:
This ordering is intentional (at least now). The ref is acquired before publishing Process->Job, so any observer of a non NULL pointer sees an object already backed by the ref owned by that field. If the CAS fails, the ref is released. Clarified in the comment.

wnsntgs committed:
[NTOS:PS] Address style nits in job implementation

tkreuzer commented on ntoskrnl/ps/job.c:
```diff
@@ -74,479 +78,2154 @@ ULONG PspJobInfoAlign[] =
+ */
+static
+NTSTATUS
+PspQueryLimitInformation(
```
I know this is a function that is local to the `job.c` file, but I'd still prefer a more specific name like `PspQueryJobLimitInformation`. Same for `PspQueryBasicAccountingInfo`.

wnsntgs replied:
Done

wnsntgs committed:
[NTOS:PS] Simplify job process enumeration

tkreuzer commented on ntoskrnl/ps/job.c:
Can be removed

tkreuzer commented on ntoskrnl/ps/job.c:
Can be removed

tkreuzer commented on ntoskrnl/ps/job.c:
`_SEH2_YIELD` is no longer required, a simple return is enough.

tkreuzer commented on ntoskrnl/ps/job.c:
I'd prefer an early return with a `DPRINT1` here.

MishaProductions commented:
`D:\a\reactos\reactos\src\ntoskrnl\ps\psmgr.c(437): error C4113: 'void (__cdecl *)(PEPROCESS,PVOID,ACCESS_MASK,ULONG,ULONG)' differs in parameter lists from 'OB_CLOSE_METHOD'`

wnsntgs commented:
> `D:\a\reactos\reactos\src\ntoskrnl\ps\psmgr.c(437): error C4113: 'void (__cdecl *)(PEPROCESS,PVOID,ACCESS_MASK,ULONG,ULONG)' differs in parameter lists from 'OB_CLOSE_METHOD'`

Thanks, fixed

HBelusca commented on dll/win32/kernel32/client/job.c:
```diff
@@ -116,11 +116,13 @@ AssignProcessToJobObject(IN HANDLE hJob,
+    _In_opt_ HANDLE hJob,
+    _In_ JOBOBJECTINFOCLASS JobObjectInformationClass,
+    _Out_writes_bytes_to_(cbJobObjectInformationLength, *lpReturnLength) LPVOID
+    lpJobObjectInformation,
```
```suggestion
    _Out_writes_bytes_to_(cbJobObjectInformationLength, *lpReturnLength)
        LPVOID lpJobObjectInformation,
```

HBelusca commented on dll/win32/kernel32/client/job.c:
```diff
@@ -23,8 +23,8 @@
-CreateJobObjectA(IN LPSECURITY_ATTRIBUTES lpJobAttributes,
-                 IN LPCSTR lpName)
+CreateJobObjectA(_In_ LPSECURITY_ATTRIBUTES lpJobAttributes,
+                 _In_ LPCSTR lpName)
```
```suggestion
CreateJobObjectA(
    _In_ LPSECURITY_ATTRIBUTES lpJobAttributes,
    _In_ LPCSTR lpName)
```

HBelusca commented on dll/win32/kernel32/client/job.c:
```diff
@@ -35,8 +35,8 @@ CreateJobObjectA(IN LPSECURITY_ATTRIBUTES lpJobAttributes,
-CreateJobObjectW(IN LPSECURITY_ATTRIBUTES lpJobAttributes,
-                 IN LPCWSTR lpName)
+CreateJobObjectW(_In_ LPSECURITY_ATTRIBUTES lpJobAttributes,
+                 _In_ LPCWSTR lpName)
```
```suggestion
CreateJobObjectW(
    _In_ LPSECURITY_ATTRIBUTES lpJobAttributes,
    _In_ LPCWSTR lpName)
```

HBelusca commented on dll/win32/kernel32/client/job.c:
```diff
@@ -47,9 +47,9 @@ CreateJobObjectW(IN LPSECURITY_ATTRIBUTES lpJobAttributes,
-               IN LPCWSTR lpName)
+OpenJobObjectW(_In_ DWORD dwDesiredAccess,
+               _In_ BOOL bInheritHandle,
+               _In_ LPCWSTR lpName)
```
```suggestion
OpenJobObjectW(
    _In_ DWORD dwDesiredAccess,
    _In_ BOOL bInheritHandle,
    _In_ LPCWSTR lpName)
```

HBelusca commented on dll/win32/kernel32/client/job.c:
```diff
@@ -61,9 +61,9 @@ OpenJobObjectW(IN DWORD dwDesiredAccess,
-               IN LPCSTR lpName)
+OpenJobObjectA(_In_ DWORD dwDesiredAccess,
+               _In_ BOOL bInheritHandle,
+               _In_ LPCSTR lpName)
```
```suggestion
OpenJobObjectA(
    _In_ DWORD dwDesiredAccess,
    _In_ BOOL bInheritHandle,
    _In_ LPCSTR lpName)
```

HBelusca commented on dll/win32/kernel32/client/job.c:
```diff
@@ -74,9 +74,9 @@ OpenJobObjectA(IN DWORD dwDesiredAccess,
-               OUT PBOOL Result)
+IsProcessInJob(_In_ HANDLE ProcessHandle,
+               _In_opt_ HANDLE JobHandle,
+               _Out_ PBOOL Result)
```
```suggestion
IsProcessInJob(
    _In_ HANDLE ProcessHandle,
    _In_opt_ HANDLE JobHandle,
    _Out_ PBOOL Result)
```

HBelusca commented on dll/win32/kernel32/client/job.c:
```diff
@@ -96,8 +96,8 @@ IsProcessInJob(IN HANDLE ProcessHandle,
-AssignProcessToJobObject(IN HANDLE hJob,
-                         IN HANDLE hProcess)
+AssignProcessToJobObject(_In_ HANDLE hJob,
+                         _In_ HANDLE hProcess)
```
```suggestion
AssignProcessToJobObject(
    _In_ HANDLE hJob,
    _In_ HANDLE hProcess)
```

HBelusca commented on dll/win32/kernel32/client/job.c:
```diff
@@ -315,8 +318,8 @@ SetInformationJobObject(IN HANDLE hJob,
-TerminateJobObject(IN HANDLE hJob,
-                   IN UINT uExitCode)
+TerminateJobObject(_In_ HANDLE hJob,
+                   _In_ UINT uExitCode)
```
```suggestion
TerminateJobObject(
    _In_ HANDLE hJob,
    _In_ UINT uExitCode)
```

HBelusca commented on dll/win32/kernel32/client/proc.c:
```diff
@@ -3808,10 +3808,10 @@ CreateProcessInternalW(IN HANDLE hUserToken,
-    /* Check if we're going to be debugged */
-    if (dwCreationFlags & DEBUG_PROCESS)
+    /* CREATE_BREAKAWAY_FROM_JOB allows the child process to break
+       away from the job associated with the calling process */
```
```suggestion
    /* CREATE_BREAKAWAY_FROM_JOB allows the child process to break
     * away from the job associated with the calling process */
```

HBelusca commented on ntoskrnl/include/internal/ps.h:
```diff
@@ -76,6 +76,27 @@
+// Flags for JobStatus in EPROCESS
+//
+// These are based on the layout of bit fields in the Flags2 set introduced in
+// version 6.0
```
```suggestion
// These are based on the layout of bit fields in the Flags2 set
// introduced in version 6.0
```

HBelusca commented on ntoskrnl/ps/kill.c:
```diff
@@ -1120,10 +1120,22 @@ PspExitProcess(IN BOOLEAN LastThread,
+            ExEnterCriticalRegionAndAcquireResourceExclusive(&Process->Job->JobLock);
+
+            /* Check if we are part of a Job that has a completion port
+               and do I/O completion if needed */
```
```suggestion
            /* Check if we are part of a job that has a completion port
             * and do I/O completion if needed */
```

HBelusca commented on ntoskrnl/ps/process.c:
```diff
@@ -709,11 +715,36 @@ PspCreateProcess(OUT PHANDLE ProcessHandle,
+     * Attach the process to parent's job if:
+     *  a) parent exists,
+     *  b) parent has a job,
+     *  c) parent's job does NOT allow silent breakaway
```
```suggestion
     *  c) parent's job does NOT allow silent breakaway.
```

HBelusca commented on ntoskrnl/ps/process.c:
```diff
@@ -376,7 +376,10 @@ PspCreateProcess(OUT PHANDLE ProcessHandle,
     SECURITY_SUBJECT_CONTEXT SubjectContext;
     BOOLEAN NeedsPeb = FALSE;
     INITIAL_PEB InitialPeb;
+    PEJOB ParentJob;
```
```suggestion
```

HBelusca commented on ntoskrnl/ps/process.c:
```diff
@@ -709,11 +715,36 @@ PspCreateProcess(OUT PHANDLE ProcessHandle,
     {
-        /* FIXME: We need to insert this process */
-        DPRINT1("Jobs not yet supported\n");
+        ParentJob = Parent->Job;
```
```suggestion
        PEJOB ParentJob = Parent->Job;
```

HBelusca commented on sdk/include/ndk/pstypes.h:
```diff
@@ -232,6 +232,22 @@ extern POBJECT_TYPE NTSYSAPI PsJobType;
+//
+// Job message flags
+//
+
```
```suggestion
```

HBelusca commented on sdk/include/ndk/pstypes.h:
```diff
@@ -232,6 +232,22 @@ extern POBJECT_TYPE NTSYSAPI PsJobType;
 #define JOB_OBJECT_SECURITY_FILTER_TOKENS       0x0008
 
+//
+// Job message flags
```
Note that since these are consecutive numbers, they look more like an enumeration to me (and so, a given value excludes the others, and cannot combines with them).

wnsntgs replied:
I think the word "flags" is wrong here, changed to "Job Messages"

HBelusca commented on sdk/include/ndk/pstypes.h:
```diff
@@ -232,6 +232,22 @@ extern POBJECT_TYPE NTSYSAPI PsJobType;
+#define JOB_OBJECT_MSG_END_OF_PROCESS_TIME      2
+#define JOB_OBJECT_MSG_ACTIVE_PROCESS_LIMIT     3
+#define JOB_OBJECT_MSG_ACTIVE_PROCESS_ZERO      4
+#define JOB_OBJECT_MSG_NEW_PROCESS              6
```
There isn't a define for value 5 ?

wnsntgs replied:
No

HBelusca commented on ntoskrnl/ps/kill.c:
```diff
@@ -1120,10 +1120,22 @@ PspExitProcess(IN BOOLEAN LastThread,
+            if (Process->Job->CompletionPort &&
+                !FlagOn(Process->JobStatus, JOB_NOT_REALLY_ACTIVE))
+            {
+                (VOID)PspSendJobMessageLocked(Process->Job,
```
Maybe retrieve the status and do a DPRINT1 or whatever if the Status isn't a success code?

wnsntgs replied:
SendJobMessageLocked can fail under memory pressure, but ordinary job messages are best-effort notifications and their delivery is not guaranteed. Eg, see https://learn.microsoft.com/en-us/windows/win32/api/winnt/ns-winnt-jobobject_associate_completion_port ("except for limits set with the JobObjectNotificationLimitInformation information class, messages are intended only as notifications and their delivery to the completion port is not guaranteed"). Other job notifications ignore the status as well, so I'd prefer to keep the exit path consistent.

HBelusca commented on ntoskrnl/include/internal/ps.h:
```diff
@@ -371,18 +392,37 @@ PspQueryDescriptorThread(
 // Job Routines
 //
+typedef NTSTATUS
+(*PJOB_ENUMERATOR_CALLBACK)(
```
I would recommend making the callback NTAPI; what do you think about that?

wnsntgs replied:
Done

HBelusca commented on ntoskrnl/ps/job.c:
```diff
@@ -2,19 +2,23 @@
- *                  Thomas Weidenmueller <w3seek@reactos.com>
- *                  Pierre Schweitzer (pierre@reactos.org)
+ * PURPOSE:         Core functions for managing Job Objects, a kernel mechanism
+ *                  for managing multiple processes as a single unit
```
```suggestion
 *                  for managing multiple processes as a single unit.
```

HBelusca commented on ntoskrnl/ps/job.c:
```diff
@@ -74,479 +78,2076 @@ ULONG PspJobInfoAlign[] =
+    Process = CONTAINING_RECORD(Entry, EPROCESS, JobLinks);
+
+    ASSERT(Process->Job == Job);
+
```
```suggestion
    ASSERT(Process->Job == Job);
```

HBelusca commented on ntoskrnl/ps/job.c:
```diff
@@ -74,479 +78,2076 @@ ULONG PspJobInfoAlign[] =
+    Process = PspGetNextProcessInJob(Job, NULL);
+
+    /* Iterate through all processes in the job */
+    while (Process != NULL)
```
Note that this could be condensed in a more elegant for-loop:
```suggestion
    /* Iterate through all processes in the job */
    for (Process = PspGetNextProcessInJob(Job, NULL);
         Process != NULL;
         Process = PspGetNextProcessInJob(Job, Process))
```

wnsntgs replied:
Done

HBelusca commented on ntoskrnl/ps/job.c:
```diff
@@ -74,479 +78,2076 @@ ULONG PspJobInfoAlign[] =
+        }
+
+        /* Move to the next process */
+        Process = PspGetNextProcessInJob(Job, Process);
```
This goes together with the for-loop suggestion:
```suggestion
```

HBelusca commented on ntoskrnl/ps/job.c:
```diff
@@ -74,479 +78,2076 @@ ULONG PspJobInfoAlign[] =
+        {
+            /*
+             * On successful iteration, PspGetNextProcessInJob consumes this
+             * reference. On failure, it must be released explicitly.
```
```suggestion
             * On successful iteration, PspGetNextProcessInJob consumes
             * this reference. On failure, it must be released explicitly.
```

HBelusca commented on ntoskrnl/ps/job.c:
```diff
@@ -74,479 +78,2076 @@ ULONG PspJobInfoAlign[] =
+    while (Process != NULL)
+    {
+        Status = Callback(Process, Context);
+
```
```suggestion
```

HBelusca commented on ntoskrnl/ps/job.c:
```diff
@@ -74,479 +78,2076 @@ ULONG PspJobInfoAlign[] =
+        if (!NT_SUCCESS(Status))
+        {
+            break;
+        }
```
```suggestion
        if (!NT_SUCCESS(Status))
            break;
```

HBelusca commented on ntoskrnl/ps/job.c:
```diff
@@ -74,479 +78,2076 @@ ULONG PspJobInfoAlign[] =
+    ASSERT(ExIsResourceAcquiredSharedLite(&Job->JobLock) != 0 ||
+           ExIsResourceAcquiredExclusiveLite(&Job->JobLock) != 0);
+
+    Status = IoSetIoCompletion(Job->CompletionPort,
```
Suggestion:
`    return IoSetIoCompletion(...);`
without the intermediate `Status` variable.

wnsntgs replied:
Done

HBelusca commented on ntoskrnl/ps/job.c:
```diff
@@ -74,479 +78,2076 @@ ULONG PspJobInfoAlign[] =
+
+    ExReleaseResourceAndLeaveCriticalRegion(&Job->JobLock);
+
+    ExReleaseRundownProtection(&Process->RundownProtect);
```
```suggestion
    ExReleaseResourceAndLeaveCriticalRegion(&Job->JobLock);
    ExReleaseRundownProtection(&Process->RundownProtect);
```

HBelusca commented on ntoskrnl/ps/job.c:
```diff
@@ -2,19 +2,23 @@
+ *                  2018 Pierre Schweitzer (pierre@reactos.org)
+ *                  2022 Timo Kreuzer (timo.kreuzer@reactos.org)
+ *                  2024-2026 Gleb Surikov (glebs.surikovs@gmail.com)
  */
```
Replacement for the whole header:
```suggestion
/*
 * PROJECT:     ReactOS Kernel
 * LICENSE:     GPL-2.0-or-later (https://spdx.org/licenses/GPL-2.0-or-later)
 * PURPOSE:     Core functions for managing Job Objects, a kernel mechanism
 *              for managing multiple processes as a single unit.
 * COPYRIGHT:   Copyright 2004-2012 Alex Ionescu (alex@relsoft.net) (stubs)
 *              Copyright 2004-2005 Thomas Weidenmueller <w3seek@reactos.com>
 *              Copyright 2015-2016 Samuel Serapión Vega <encoded@reactos.org>
 *              Copyright 2017 Mark Jansen <mark.jansen@reactos.org>
 *              Copyright 2018 Pierre Schweitzer <pierre@reactos.org>
 *              Copyright 2022 Timo Kreuzer <timo.kreuzer@reactos.org>
 *              Copyright 2024-2026 Gleb Surikov <glebs.surikovs@gmail.com>
 */
```

wnsntgs replied:
Done

HBelusca commented on ntoskrnl/include/internal/ps.h:
```diff
@@ -76,6 +76,27 @@
+// https://www.geoffchappell.com/studies/windows/km/ntoskrnl/inc/ntos/ps/eprocess/flags2.htm
+//
+#define JOB_NOT_REALLY_ACTIVE   0x00000001
+#define ACCOUNTING_FOLDED       0x00000002
```
Wouldn't it be best to prefix these with a "JOB_" or "PS_JOB_" or something? Just a name like that could pretty much collide with a completely unrelated macro someone could define somewhere.

wnsntgs replied:
Done

HBelusca commented on ntoskrnl/ps/job.c:
```diff
@@ -764,23 +2302,29 @@ NtSetInformationJobObject (
     if (PreviousMode != KernelMode)
     {
-        ASSERT(((RequiredAlign) == 1) || ((RequiredAlign) == 2) || ((RequiredAlign) == 4) || ((RequiredAlign) == 8) || ((RequiredAlign) == 16));
+        ASSERT(((RequiredAlign) == 1) ||
```
Since you've modified the formatting, you may also remove the redundant parentheses around "RequiredAlign".

wnsntgs replied:
Done

HBelusca commented on ntoskrnl/ps/job.c:
```diff
@@ -797,23 +2341,27 @@ NtSetInformationJobObject (
                                        PsJobType,
                                        PreviousMode,
-                                       (PVOID*)&Job,
+                                       (PVOID *)&Job,
```
Not necessary

HBelusca commented on ntoskrnl/ps/job.c:
```diff
@@ -822,110 +2370,84 @@ NtSetInformationJobObject (
 
     /* And set the information */
     KeEnterGuardedRegionThread(CurrentThread);
```
Naive question: what's the reason for the guarded region? Does it play a role similar to "locking the thread" in order to modify the job info?

wnsntgs replied:
It disables kernel APC delivery for the duration of the set operation. This is established once at the NtSetInformationJobObject level so helpers that require APCs to be disabled can use the corresponding unsafe operations without entering/leaving another guarded region themselves. I clarified the comment.

HBelusca replied:
Thanks

HBelusca commented on ntoskrnl/ps/job.c:
```diff
@@ -822,110 +2370,84 @@ NtSetInformationJobObject (
+        _SEH2_TRY
+        {
+            /* If asking for extending limits */
+            if (JobInformationClass == JobObjectExtendedLimitInformation)
```
```suggestion
            if (IsExtendedLimit)
```

HBelusca commented on ntoskrnl/ps/job.c:
See other comment

HBelusca commented on ntoskrnl/ps/job.c:
```suggestion
    if (!NT_SUCCESS(Status))
```
and in other similar places.

HBelusca commented on ntoskrnl/ps/job.c:
for example there...

HBelusca commented on ntoskrnl/ps/job.c:
There's no existing defines for these scheduling classes (and for this "5' in particular)?

wnsntgs replied:
No, but added a define for the default class

HBelusca commented on ntoskrnl/ps/job.c:
```diff
@@ -74,479 +77,2065 @@ ULONG PspJobInfoAlign[] =
+                                              PspAssociateCompletionPortCallback,
+                                              Job);
+
+    ASSERT(NT_SUCCESS(Status));
```
Why not just returning the actual status code below? (I don't fully understand the reason exposed in the associated comment)

wnsntgs replied:
Returning status would be equivalent with the current callback, since it always returns success. I think keeping the explicit success is preferable though, as by the time we enumerate, the IOCP association has already been committed, while the initial notifications are best-effort. Propagating a later notification/enumeration failure would report failure to the caller even though the job remains associated with the port. I clarified the comment a bit.

HBelusca commented on ntoskrnl/ps/job.c:
```diff
@@ -74,479 +77,2065 @@ ULONG PspJobInfoAlign[] =
+        JOB_OBJECT_LIMIT_PRESERVE_JOB_TIME |
+        JOB_OBJECT_LIMIT_SCHEDULING_CLASS;
+
+    const ULONG AllowedExtendedFlags = AllowedBasicFlags |
```
Or make defines?

wnsntgs replied:
Done

HBelusca commented on ntoskrnl/ps/job.c:
```diff
@@ -74,479 +77,2065 @@ ULONG PspJobInfoAlign[] =
+ * @returns
+ *     STATUS_SUCCESS if the job limits are successfully set.
+ *     Otherwise, an appropriate NTSTATUS error code.
+ *
```
```suggestion
```
